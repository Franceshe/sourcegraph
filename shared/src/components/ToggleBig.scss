.toggle-big {
    background: none;
    border: none;
    cursor: pointer;
    display: inline-block;
    outline: none !important;
    padding: 0;
    position: relative;
    width: 6.625rem;
}

.toggle-big__bar {
    background-color: var(--secondary);

    border-radius: 2rem;

    border-radius: 1rem;
    cursor: pointer;

    top: 2px;
    left: 0;

    height: 2rem;
    width: 100%;

    position: absolute;
}

.toggle-big__bar--active {
    background-color: rgba(28, 126, 214, 0.2);
}

.toggle-big__knob {
    background-color: var(--text-muted);
    opacity: 0.8;
    border-radius: 0.75rem;

    display: block;

    height: 1.5rem;
    width: 1.5rem;
    margin-top: 0.375rem;
    left: 0.25rem;

    position: relative;
}

.toggle-big__knob--active {
    background-color: var(--primary);
    transform: translate3d(4.625rem, 0, 0);
}

.toggle-big__knob--disabled {
    background-color: rgba(43, 55, 80, 0.2);

    &__active {
        background-color: rgba(28, 126, 214, 0.2);
    }
}

.toggle-big__knob--icon {
    position: relative;
    margin-top: 0.375rem;
    left: 0.4rem;
    fill: white;

    &__disabled {
        left: -4.4rem;
        fill: #e1e6ef;
    }
}

.toggle-big__text {
    color: var(--primary);
    position: relative;
    font-size: 0.875rem;
    margin-top: 0.375rem;
    right: 0.4375rem;

    &--disabled {
        color: var(--text-muted);
        right: -0.6875rem;
    }
}
