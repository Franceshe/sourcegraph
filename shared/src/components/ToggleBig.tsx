import * as React from 'react'
import Close from 'mdi-react/CloseIcon'
import Check from 'mdi-react/CheckIcon'
import classnames from 'classnames'

interface Props {
    /** The initial value. */
    value?: boolean

    /** The DOM ID of the element. */
    id?: string

    /**
     * Called when the user changes the input's value.
     */
    onToggle?: (value: boolean) => void

    /**
     * Called when the user hovers over the toggle.
     */
    onHover?: (value: boolean) => void

    /** The title attribute (tooltip). */
    title?: string

    disabled?: boolean
    tabIndex?: number
    className?: string
}

/** A big toggle switch input component. */
export const ToggleBig: React.FunctionComponent<Props> = ({
    disabled,
    className,
    id,
    title,
    value,
    tabIndex,
    onToggle,
    onHover,
}) => {
    function onClick(): void {
        if (!disabled && onToggle) {
            onToggle(!value)
        }
    }

    function onMouseOver(): void {
        if (onHover) {
            onHover(!value)
        }
    }

    return (
        <button
            type="button"
            className={classnames('toggle-big', className, {
                'toggle-big--disabled': !!disabled,
            })}
            id={id}
            title={title}
            value={value ? 1 : 0}
            onClick={onClick}
            tabIndex={tabIndex}
            onMouseOver={onMouseOver}
            disabled={!!disabled}
        >
            <span className="toggle-big__container">
                <span
                    className={`toggle-big__bar ${value ? 'toggle-big__bar--active' : ''} ${
                        disabled ? 'toggle-big__bar--disabled' : ''
                    }`}
                />
                <span className={`toggle-big__knob ${value ? 'toggle-big__knob--active' : ''}`}>
                    {value ? (
                        <Check size={16} className="toggle-big__icon" />
                    ) : (
                        <Close size={16} className="toggle-big__icon--disabled" />
                    )}
                </span>
                <span className={`toggle-big__text ${!value ? 'toggle-big__text--disabled' : ''}`}>
                    {value ? 'Enabled' : 'Disabled'}
                </span>
            </span>
        </button>
    )
}
