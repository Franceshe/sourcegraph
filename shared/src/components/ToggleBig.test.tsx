import React from 'react'
import renderer from 'react-test-renderer'
import { ToggleBig } from './ToggleBig'

describe('Toggle', () => {
    test('value is false', () => {
        const component = renderer.create(<ToggleBig value={false} />)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })

    test('value is true', () => {
        const component = renderer.create(<ToggleBig value={true} />)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })

    test('disabled', () => {
        const component = renderer.create(<ToggleBig disabled={true} />)
        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()

        // Clicking while disabled is a noop.
        tree!.props.onClick({ preventDefault: () => undefined, currentTarget: { blur: () => undefined } })
        tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })

    test('className', () => expect(renderer.create(<ToggleBig className="c" />).toJSON()).toMatchSnapshot())
})
