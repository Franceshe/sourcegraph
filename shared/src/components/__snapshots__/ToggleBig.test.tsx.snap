// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle className 1`] = `
<button
  className="toggle-big  c"
  onClick={[Function]}
  type="button"
  value={0}
>
  <span
    className="toggle-big__container"
  >
    <span
      className="toggle-big__bar  "
    />
    <span
      className="toggle-big__knob "
    />
    <span
      className="toggle-big__text toggle-big__text--disabled"
    >
      Disabled
    </span>
    <svg
      className="mdi-icon toggle-big__icon toggle-big__icon--disabled"
      fill="currentColor"
      height={16}
      viewBox="0 0 24 24"
      width={16}
    >
      <path
        d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
      />
    </svg>
  </span>
</button>
`;

exports[`Toggle disabled 1`] = `
<button
  className="toggle-big toggle-big__disabled "
  onClick={[Function]}
  type="button"
  value={0}
>
  <span
    className="toggle-big__container"
  >
    <span
      className="toggle-big__bar  toggle-big__bar--disabled"
    />
    <span
      className="toggle-big__knob "
    />
    <span
      className="toggle-big__text toggle-big__text--disabled"
    >
      Disabled
    </span>
    <svg
      className="mdi-icon toggle-big__icon toggle-big__icon--disabled"
      fill="currentColor"
      height={16}
      viewBox="0 0 24 24"
      width={16}
    >
      <path
        d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
      />
    </svg>
  </span>
</button>
`;

exports[`Toggle disabled 2`] = `
<button
  className="toggle-big toggle-big__disabled "
  onClick={[Function]}
  type="button"
  value={0}
>
  <span
    className="toggle-big__container"
  >
    <span
      className="toggle-big__bar  toggle-big__bar--disabled"
    />
    <span
      className="toggle-big__knob "
    />
    <span
      className="toggle-big__text toggle-big__text--disabled"
    >
      Disabled
    </span>
    <svg
      className="mdi-icon toggle-big__icon toggle-big__icon--disabled"
      fill="currentColor"
      height={16}
      viewBox="0 0 24 24"
      width={16}
    >
      <path
        d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
      />
    </svg>
  </span>
</button>
`;

exports[`Toggle value is false 1`] = `
<button
  className="toggle-big  "
  onClick={[Function]}
  type="button"
  value={0}
>
  <span
    className="toggle-big__container"
  >
    <span
      className="toggle-big__bar  "
    />
    <span
      className="toggle-big__knob "
    />
    <span
      className="toggle-big__text toggle-big__text--disabled"
    >
      Disabled
    </span>
    <svg
      className="mdi-icon toggle-big__icon toggle-big__icon--disabled"
      fill="currentColor"
      height={16}
      viewBox="0 0 24 24"
      width={16}
    >
      <path
        d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
      />
    </svg>
  </span>
</button>
`;

exports[`Toggle value is true 1`] = `
<button
  className="toggle-big  "
  onClick={[Function]}
  type="button"
  value={1}
>
  <span
    className="toggle-big__container"
  >
    <span
      className="toggle-big__bar toggle-big__bar--active "
    />
    <span
      className="toggle-big__knob toggle-big__knob--active"
    />
    <span
      className="toggle-big__text "
    >
      Enabled
    </span>
    <svg
      className="mdi-icon toggle-big__icon"
      fill="currentColor"
      height={16}
      viewBox="0 0 24 24"
      width={16}
    >
      <path
        d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
      />
    </svg>
  </span>
</button>
`;
